src/data/nat/digits.lean

```diff
@@ -444,3 +444,28 @@ end
 
 lemma le_digits_len_le (b n m : ℕ) (h : n ≤ m) : (digits b n).length ≤ (digits b m).length :=
 monotone_of_monotone_nat (digits_len_le_digits_len_succ b) h
+
+lemma digits_ge_base_pow_len (b m : ℕ) : m ≠ 0 → (b + 2) ^ ((digits (b + 2) m).length) ≤ (b + 2) * m :=
+begin
```

Comment:
I think an easier proof would be similar to the proof for `lt_base_pow_length_digits`: first do it for a list of digits, and then  use it to prove it for a natural number.

Reply:
Hi. I added a nonprime version w/the hypothesis 2 <= b. Unfortunately, I am having some difficulties w/the list version since it doesn't work for 0 or list.nil...

Reply:
I'm not sure that it would be easier to go through the list of digits tbh. The `m ≠ 0` assumption here becomes `l ≠ []` for lists, and the induction on lists might become much more complex. When I wrote `lt_base_pow_length_digits`, it's because `of_digits_lt_base_pow_length`'s induction is fairly simple. That might not be the case here.

I think proving it via lists requires the fact that `l ≠ [] -> 0 < l.last _` (at least when I attempted this proof), which isn't in mathlib yet, and I'm not sure it would be trivial to prove.

Reply:
Thanks.

Here is how I would do the other lemma:

```lean
@[simp] lemma of_digits_singleton {b n : ℕ} : of_digits b [n] = n := by simp [of_digits]

open list
lemma pow_length_le_mul_of_digits {b : ℕ} {l : list ℕ} (hl : l ≠ []) (h2l : l.last hl ≠ 0) :
  (b + 2) ^ l.length ≤ (b + 2) * of_digits (b+2) l :=
begin
  rw [← init_append_last hl],
  simp only [length_append, length, zero_add, length_init, of_digits_append, length_init,
    of_digits_singleton, add_comm (l.length - 1), nat.pow_add, nat.pow_one],
  apply nat.mul_le_mul_left,
  refine le_trans _ (nat.le_add_left _ _),
  have : 1 ≤ l.last hl, { rwa [nat.succ_le_iff, nat.pos_iff_ne_zero] },
  convert nat.mul_le_mul_left _ this, rw [mul_one]
end
```

Reply:
Note: this uses a lemma about list that was added today(!), so you might need to update mathlib before this proof works for you.

Reply:
Is this something that's blocking the PR from being merged? In my opinion the method going through a list of digits will probably be a nicer proof, however there is still missing API, especially to do with `(digits b m).last`, which doesn't seem trivial to do from what I can tell (I've ended up in `motive is not type correct` hell when trying to prove it)

Would it be better to have this PR as-is, and then maybe refactor the proof once we have `0 < (digits _ _).last` in mathlib?

Reply:
I guess we can merge the PR now.
Normally it is good practice to first build the basic API for functions before proving more complicated things about them. I'm very surprised that `0 < (digits b m).last` is still missing, given that it is mentioned in the docstring of `digits`.
That way you can make your proofs much shorter. With API, you don't need to do induction to show this fact anymore.

I'm happy to merge this if another reviewer agrees it is ready.

Reply:
I think I'm going to try to look at `0 < (digits b m).last` then (although it's unlikely I'll get anywhere...). Thanks so much for all the feedback!

Reply:
For what it's worth, here's what I have so far. I've managed to do the cases for `b = 0` and `b = 1`, but for base `b + 2`, I think strong induction will be necessary. However I'm kind of stuck on `motive is not type correct`, and I'm not sure how it could be avoided tbh.

https://gist.github.com/shingtaklam1324/375281a96af94cea940d49634f4af0ca